### PR TITLE
list of exclude vars

### PR DESCRIPTION
**Pull request by cboettig**
@eitsupi @Pit-Storm from #188 looks like this sucks up PASSWORD too, which ought to be excluded. vectorized the logic here (I think).

### CODE REVIEW

**Comment by eitsupi** [comment]
I'm not familiar with shellscript so I don't understand how this change works, but it looks great !

The fact that I could see the PASSWORD was bothering me too, but I couldn't determine if it was a problem.
(Users enter the same string as PASSWORD when they log in to RStudio.)

What I was wondering is that every time I reboot, `/init` will be executed and environment variables will be written to `${R_HOME}/etc/Renviron.site`, which will increase the file size, but I don't know how to stop this.

**Comment by eitsupi** [comment]
I hadn't noticed this before, but I think we can read the PASSWORD even if it doesn't exist in the environment variable by looking in `/var/run/s6/container_environment/`.

I tried to run it from bash in RStudio in `rocker/rstudio:4.0.5`, where no updates #186 have been applied.

```shell
rstudio@31630821fdce:~$ ls /var/run/s6/container_environment/
CRAN  CWD  HOME  HOSTNAME  LANG  LC_ALL  PASSWORD  PATH  R_HOME  RSTUDIO_VERSION  R_VERSION  S6_VERSION  TERM  TZ
rstudio@31630821fdce:~$ cat /var/run/s6/container_environment/PASSWORD
pass
```

**Comment by cboettig** [comment]
> What I was wondering is that every time I reboot, /init will be executed and environment variables will be written to ${R_HOME}/etc/Renviron.site, which will increase the file size, but I don't know how to stop this.

We could add a bit of extra bash logic to handle that.  I think `sed -i` might be the natural choice(?), e.g. (untested!)

```bash
sed -i "/$VAR/d" $R_HOME/etc/Renviron.site
```

before appending that `$VAR`.

(Arguably we should do so in the docker scripts as well, as long as we're careful to do it in bash and not sh scripts).  


Good point about `cat /var/run/s6/container_environment/PASSWORD` too.  Like you say,  this doesn't really matter in a single-user setup, since the user is only seeing their own password then anyhow.  However, in multi-user environments that is perhaps not ideal (though again, probably not a real issue if no-one is using the `rstudio` user account and that account has not been granted elevated privileges).  In any event, this could be altered by changing permissions on the file, e.g. something like

```
chmod 600 /var/run/s6/container_environment/*
```

to make the files readable only to root.

**Comment by cboettig** [comment]
ok, penciled both of these into useconf.sh but haven't tested yet.

File: scripts/userconf.sh [hunk 1]
```diff
@@ -107,12 +107,16 @@ fi
...
 ## Set our dynamic variables in Renviron.site to be reflected by RStudio
 for file in /var/run/s6/container_environment/*
 do
-  if [ "${file##*/}" != "HOME" ]; then
-    echo "${file##*/}=$(cat $file)" >> ${R_HOME}/etc/Renviron.site
-  fi
+  ## Remove variable if it already exists, to avoid repeated declaration on container-restart:
+  sed -i "/$file/d" ${R_HOME}/etc/Renviron.site
+  ## Only add non-excluded variables (space-sep list above) to Renviron.site
+  [[ $exclude_vars =~ (^| )$file($| ) ]] && echo "${file##*/}=$(cat $file)" >> ${R_HOME}/etc/Renviron.site || echo "skipping $file"
```
Comment (eitsupi) [inline scripts/userconf.sh:119]: It doesn't seem to be working as expected.
It seems to work well if you write it like this

```shell
for file in /var/run/s6/container_environment/*
do
  regex="(^| )${file##*/}($| )"
  [[ ! $exclude_vars =~ $regex ]] && echo "${file##*/}=$(cat $file)"
done
```

**Review by eitsupi — CHANGES_REQUESTED** [review]

File: scripts/userconf.sh [hunk 2]
```diff
@@ -107,12 +107,16 @@ fi
...
+
+exclude_vars="HOME PASSWORD"
 ## Set our dynamic variables in Renviron.site to be reflected by RStudio
 for file in /var/run/s6/container_environment/*
 do
-  if [ "${file##*/}" != "HOME" ]; then
-    echo "${file##*/}=$(cat $file)" >> ${R_HOME}/etc/Renviron.site
-  fi
+  ## Remove variable if it already exists, to avoid repeated declaration on container-restart:
+  sed -i "/$file/d" ${R_HOME}/etc/Renviron.site
```
Comment (eitsupi) [inline scripts/userconf.sh:117]: It seems to work if we change it like follows?

```shell
sed -i "/^${file##*/}/d" ${R_HOME}/etc/Renviron.site
```

**Review by eitsupi — CHANGES_REQUESTED** [review]

**Comment by eitsupi** [comment]
I copied the following slightly modified script as commented above to `rocker/rstudio` and played with it in RStudio, and it seems to work well.

```shell
exclude_vars="HOME PASSWORD"
for file in /var/run/s6/container_environment/*
do
  sed -i "/^${file##*/}=/d" ${R_HOME}/etc/Renviron.site
  regex="(^| )${file##*/}($| )"
  [[ ! $exclude_vars =~ $regex ]] && echo "${file##*/}=$(cat $file)" >> ${R_HOME}/etc/Renviron.site || echo "skipping $file"
done
```

The following process for writing to Renviron.site seems redundant and can be removed.

https://github.com/rocker-org/rocker-versioned2/blob/6eb2536a4be9a433948c3715ea421ec876f025fc/scripts/userconf.sh#L13-L14
https://github.com/rocker-org/rocker-versioned2/blob/6eb2536a4be9a433948c3715ea421ec876f025fc/scripts/userconf.sh#L107-L109
https://github.com/rocker-org/rocker-versioned2/blob/6eb2536a4be9a433948c3715ea421ec876f025fc/scripts/userconf.sh#L123-L124

Perhaps we need to check the behavior if we install shinyserver as well?

File: scripts/userconf.sh (hunk 2 above)
Comment (cboettig) [reply scripts/userconf.sh]: right, thanks for testing. updated.

**Comment by cboettig** [comment]
@eitsupi excellent, think I got the edits mentioned above in now.  Maybe give it one more look if you get a chance before we merge?

**Comment by eitsupi** [comment]
@cboettig Thanks !
I copied it to `rocker/rstudio` and ran it as I did in #189, and I think it is working as intended. I post the log.

I later realized that this edit eliminates the behavior of `$LANG` being copied to `$LC_ALL` (you can see at the end of the log that `$LC_ALL` is not overwritten by `$LANG`).
But judging from comment in #186 by @Pit-Storm, I think this is due to the typo in #186 and the behavior of not being overwritten seems to be normal.
I don't have a solid understanding of the locale, so I apologize if I'm wrong.

<details>
<summary>Log</summary>
<pre>
<code>
$ docker run --rm -it -e TZ=Asia/Tokyo -p 8787:8787 -e PASSWORD=foobar rockertest bash
root@3beee71e9b37:/# cat ${R_HOME}/etc/Renviron.site
R_LIBS=${R_LIBS-'/usr/local/lib/R/site-library:/usr/local/lib/R/library'}
PATH=/usr/lib/rstudio-server/bin:/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
root@3beee71e9b37:/# /init
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] userconf: executing...
skipping /var/run/s6/container_environment/HOME
skipping /var/run/s6/container_environment/PASSWORD
[cont-init.d] userconf: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
25 Jun 2021 16:24:43 [rsession-rstudio] WARNING FileLock classes not yet initialized; please call 'FileLock::initialize()' and 'FileLock::cleanUp()' as appropriate; LOGGED FROM: static bool rstudio::core::FileLock::verifyInitialized() src/cpp/core/file_lock/FileLock.cpp:113
^C[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
s6-svwait: fatal: timed out
[s6-finish] sending all processes the TERM signal.
Hangup
root@3beee71e9b37:/# cat ${R_HOME}/etc/Renviron.site
R_LIBS=${R_LIBS-'/usr/local/lib/R/site-library:/usr/local/lib/R/library'}
_=/init
CRAN=https://packagemanager.rstudio.com/all/__linux__/focal/latest
CWD=/
HOSTNAME=3beee71e9b37
LANG=en_US.UTF-8
LC_ALL=en_US.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
PATH=/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
R_HOME=/usr/local/lib/R
RSTUDIO_VERSION=1.4.1717
R_VERSION=4.1.0
S6_VERSION=v2.1.0.2
SHLVL=1
TERM=xterm
TZ=Asia/Tokyo
root@3beee71e9b37:/# /init
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] userconf: executing...
skipping /var/run/s6/container_environment/HOME
skipping /var/run/s6/container_environment/PASSWORD
[cont-init.d] userconf: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
^C[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
s6-svwait: fatal: timed out
[s6-finish] sending all processes the TERM signal.
Hangup
root@3beee71e9b37:/# cat ${R_HOME}/etc/Renviron.site
R_LIBS=${R_LIBS-'/usr/local/lib/R/site-library:/usr/local/lib/R/library'}
_=/init
CRAN=https://packagemanager.rstudio.com/all/__linux__/focal/latest
CWD=/
HOSTNAME=3beee71e9b37
LANG=en_US.UTF-8
LC_ALL=en_US.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
PATH=/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
R_HOME=/usr/local/lib/R
RSTUDIO_VERSION=1.4.1717
R_VERSION=4.1.0
S6_VERSION=v2.1.0.2
SHLVL=1
TERM=xterm
TZ=Asia/Tokyo
root@3beee71e9b37:/# ls -l /var/run/s6/container_environment/
total 72
-rw------- 1 root root    5 Jun 26 01:28 _
-rw------- 1 root root   61 Jun 26 01:28 CRAN
-rw------- 1 root root    1 Jun 26 01:28 CWD
-rw------- 1 root root    5 Jun 26 01:28 HOME
-rw------- 1 root root   12 Jun 26 01:28 HOSTNAME
-rw------- 1 root root   11 Jun 26 01:28 LANG
-rw------- 1 root root   11 Jun 26 01:28 LC_ALL
-rw------- 1 root root 1495 Jun 26 01:28 LS_COLORS
-rw------- 1 root root    6 Jun 26 01:28 PASSWORD
-rw------- 1 root root   88 Jun 26 01:28 PATH
-rw------- 1 root root    1 Jun 26 01:28 PWD
-rw------- 1 root root   16 Jun 26 01:28 R_HOME
-rw------- 1 root root    8 Jun 26 01:28 RSTUDIO_VERSION
-rw------- 1 root root    5 Jun 26 01:28 R_VERSION
-rw------- 1 root root    8 Jun 26 01:28 S6_VERSION
-rw------- 1 root root    1 Jun 26 01:28 SHLVL
-rw------- 1 root root    5 Jun 26 01:28 TERM
-rw------- 1 root root   10 Jun 26 01:28 TZ
root@3beee71e9b37:/# export LANG=ja_JP.UTF-8
root@3beee71e9b37:/# /init
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] userconf: executing...
skipping /var/run/s6/container_environment/HOME
skipping /var/run/s6/container_environment/PASSWORD
Generating locales (this might take a while)...
  ja_JP.UTF-8... done
Generation complete.
[cont-init.d] userconf: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
^C[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
s6-svwait: fatal: timed out
[s6-finish] sending all processes the TERM signal.
Hangup
root@3beee71e9b37:/# cat ${R_HOME}/etc/Renviron.site
R_LIBS=${R_LIBS-'/usr/local/lib/R/site-library:/usr/local/lib/R/library'}
_=/init
CRAN=https://packagemanager.rstudio.com/all/__linux__/focal/latest
CWD=/
HOSTNAME=3beee71e9b37
LANG=ja_JP.UTF-8
LC_ALL=en_US.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
PATH=/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
R_HOME=/usr/local/lib/R
RSTUDIO_VERSION=1.4.1717
R_VERSION=4.1.0
S6_VERSION=v2.1.0.2
SHLVL=1
TERM=xterm
TZ=Asia/Tokyo
</code>
</pre>
</details>

**Review by eitsupi — APPROVED** [review]

**Pull request merged**